File: Java/lab19_Observer/task_2_1 – solution/src/WeatherDataCenter.java
import java.util.ArrayList;
import java.util.List;

public class WeatherDataCenter {

    final private WeatherReader weatherReader;
    final private List<Subscriber> subscribers = new ArrayList<>();
    public void subscribe(Subscriber subscriber){
        subscribers.add(subscriber);
    }
    public void unsubscribe(Subscriber subscriber){
        subscribers.remove(subscriber);
    }
    public void notifySubscribers(WeatherEvent event){
        for (Subscriber subscriber : subscribers) {
            subscriber.update(this, event);
        }
    }


    public WeatherDataCenter() {
        weatherReader = new WeatherReader("resources/input01.txt");
    }

    public boolean update() {
        WeatherEvent weatherEvent = weatherReader.getEvent();
        if (weatherEvent == null){
            return false;
        }

        if (weatherEvent.getLocation().equals("end")){
            return true;
        }

        notifySubscribers(weatherEvent);

        return false;
    }
}

File: Cpp/lab19_Observer/task_2_1 - solution/src/weather_data_center.cpp
#include <iostream>
#include <algorithm>

#include "weather_data_center.hpp"


std::string path = "resources/input01.txt";

WeatherDataCenter::WeatherDataCenter()
: weatherReader(new WeatherReader(path))
{}

WeatherDataCenter::~WeatherDataCenter()
{
  delete weatherReader;
}

void WeatherDataCenter::subscribe(Subscriber* subscriber)
{
  this->subscribers.push_back(subscriber);
}

void WeatherDataCenter::unsubscribe(Subscriber* subscriber)
{
  std::remove(this->subscribers.begin(), this->subscribers.end(), subscriber);
}

void WeatherDataCenter::notifySubscribers(WeatherEvent* event)
{
  for (auto& subscriber : this->subscribers)
  {
    subscriber->update(this, event);
  }
}

bool WeatherDataCenter::update()
{
  WeatherEvent* weatherEvent = this->weatherReader->getEvent();

  if (weatherEvent == nullptr)
  {
    return false;
  }

  if (weatherEvent->getLocation() == "end")
  {
    delete weatherEvent;
    return true;
  }

  this->notifySubscribers(weatherEvent);

  delete weatherEvent;
  return false;
}
